external/yolo_detector/yolo/utils.py

```
@@ -0,0 +1,134 @@
+
+    name, ext = os.path.splitext(path)
+    new_file = "{}_{}{}".format(name, sha256[:8], ext)
+    torch.save(weights, new_file)
+    
+    
+class TextArea:
+    def __init__(self):
+        self.buffer = []
+    
+    def write(self, s):
+        self.buffer.append(s)
+        
+    def __str__(self):
+        return "".join(self.buffer)
+
+    def get_AP(self):
+        txt = str(self)
+        values = re.findall(r"(\d{3})\n", txt)
+        if len(values) > 0:
+            values = [int(v) / 10 for v in values]
+            result = {"bbox AP": values[0]} if values else {}
+            return result
+        else:
+            return txt
+    
+    
+class Meter:
+    def __init__(self, name):
+        self.name = name
+        self.reset()
+
+    def reset(self):
+        self.val = 0
+        self.avg = 0
+        self.sum = 0
+        self.count = 0
+
+    def update(self, val, n=1):
+        self.val = val
+        self.sum += val * n
+        self.count += n
+        self.avg = self.sum / self.count
+
+    def __str__(self):
+        fmtstr = "{name}:sum={sum:.2f}, avg={avg:.4f}, count={count}"
+        return fmtstr.format(**self.__dict__)
+    
+
+class ModelEMA:
+    def __init__(self, model, decay=0.9999):
+        self.ema = copy.deepcopy(model)
+        self.ema.eval()
+        self.updates = 0
+        self.decay = lambda x: decay * (1 - math.exp(-x / 2000))
+        self.distributed = isinstance(model, torch.nn.parallel.DistributedDataParallel)
+        
+        for p in self.ema.parameters():
+            p.requires_grad_(False)
+
+    def update(self, model):
+        self.updates += 1
+        d = self.decay(self.updates)
+        with torch.no_grad():
+            if self.distributed:
+                msd, esd = model.module.state_dict(), self.ema.module.state_dict()
+            else:
+                msd, esd = model.state_dict(), self.ema.state_dict()
+
+            for k, v in esd.items():
+                if v.dtype.is_floating_point:
```

if v is not a floating-point tensor (e.g., int types or buffers), it’s silently skipped this could cause EMA corruption.

quick fix:
```python
for k, v in esd.items():
    if not msd[k].dtype.is_floating_point:
        continue  # or log
    v *= d
    v += (1. - d) * msd[k].detach()

```